src/lib/substrate/substrate-context.tsx

```diff
@@ -0,0 +1,415 @@
+import { createInterBtcApi } from '@interlay/interbtc-api';
+import { ApiPromise } from '@polkadot/api';
+import { web3Accounts, web3Enable, web3FromAddress } from '@polkadot/extension-dapp';
+import { InjectedAccountWithMeta, InjectedExtension } from '@polkadot/extension-inject/types';
+import { KeyringPair, KeyringPair$Meta } from '@polkadot/keyring/types';
+import { TypeRegistry } from '@polkadot/types/create';
+import jsonrpc from '@polkadot/types/interfaces/jsonrpc';
+import type { DefinitionRpcExt } from '@polkadot/types/types';
+import { keyring } from '@polkadot/ui-keyring';
+import { Keyring } from '@polkadot/ui-keyring/Keyring';
+import { isTestChain } from '@polkadot/util';
+import * as React from 'react';
+import { useLocalStorage } from 'react-use';
+
+import { APP_NAME } from '@/config/relay-chains';
+import { SELECTED_ACCOUNT_LOCAL_STORAGE_KEY } from '@/config/wallets';
+
+import * as constants from '../../constants';
+
+const UNABLE_TO_INITIALISE_OPTIONS_MORE_THAN_ONCE = 'Unable to initialise options more than once';
+
+enum KeyringStatus {
+  Idle = 'IDLE',
+  Loading = 'LOADING',
+  Ready = 'READY',
+  Error = 'ERROR'
+}
+
+enum ApiStatus {
+  Idle = 'IDLE',
+  ConnectInit = 'CONNECT_INIT',
+  Connecting = 'CONNECTING',
+  Ready = 'READY',
+  Error = 'ERROR',
+  Disconnected = 'DISCONNECTED'
+}
+
+enum ActionType {
+  ConnectInit = 'CONNECT_INIT',
+  Connect = 'CONNECT',
+  ConnectSuccess = 'CONNECT_SUCCESS',
+  ConnectError = 'CONNECT_ERROR',
+  ConnectFail = 'CONNECT_FAIL',
+  SetKeyringLoading = 'SET_KEYRING_LOADING',
+  SetKeyringReady = 'SET_KEYRING_READY',
+  SetKeyringError = 'SET_KEYRING_ERROR',
+  SetSelectedAccount = 'SET_SELECTED_ACCOUNT',
+  SetAccounts = 'SET_ACCOUNTS',
+  SetExtensions = 'SET_EXTENSIONS'
+}
+
+interface APIError extends Error {
+  target: {
+    url: string;
+  };
+}
+
+interface KeyringPairMeta extends KeyringPair$Meta {
+  name: string;
+}
+
+interface _KeyringPair extends KeyringPair {
```

Comment:
When I suggested this approach, I wasn't expecting that you would use the name I provided. Instead could name the imported `KeyringPair` something else and keep our side with the name `KeyringPair`

Reply:
Also this file really needs to be broken down. Its hard to review it like this.

Reply:
@danielsimao
> Also this file really needs to be broken down. Its hard to review it like this.
Oh, sorry. No worries. It was broken down for sure. I did not update you.

Reply:
> When I suggested this approach, I wasn't expecting that you would use the name I provided. Instead could name the imported KeyringPair something else and keep our side with the name KeyringPair

Done!